pkg/region/cache.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2021 Outreach Corporation. All Rights Reserved.
+
+// Description: This file contains the read through cache implementation for
+// this package.
+package region
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	// RegionCachePath is the $HOME/<RegionCachePath> location of the regionpicker cache
+	RegionCachePath = filepath.Join(".outreach", ".cache", "box")
+	// RegionCacheFile is the name of the regionpicker cache file
+	RegionCacheFile = "regions.json"
+)
+
+// cache is the package specific cache instance used. This is a global variable to
+// ensure that this is go-routine safe.
+var cache = &cacheStore{}
+
+// cacheEntry is information on a cloud's cache entry
+type cacheEntry struct {
+	// Duration is how long it took to talk to this region at LastUpdatedAt
+	Duration time.Duration `json:"duration"`
+
+	// ExpiresAt is when this cache entry expires. UTC.
+	ExpiresAt time.Time `json:"expires_at"`
+}
+
+type cacheStore struct {
+	// Clouds is a map of cloudName to cache entries
+	// cloud -> region -> cacheEntry
+	Clouds map[CloudName]map[Name]cacheEntry `json:"clouds"`
+
+	mu sync.Mutex
+}
+
+// Get returns the duration of a cloud/region pairing, if it exists.
+// Otherwise ok is returned as false and time.Duration is it's zero-value.
+func (c *cacheStore) Get(cloud CloudName, r Name) (time.Duration, bool) {
+	if c.Clouds == nil {
+		c.load()
+	}
+	c.ensureKey(cloud, r)
+
+	v, ok := c.Clouds[cloud][r]
+	if !ok {
+		return time.Duration(0), ok
+	}
+	return v.Duration, ok
+}
+
+// ensureKey ensures that a key can be properly accessed in the underlying cache
+func (c *cacheStore) ensureKey(cloud CloudName, r Name) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if _, ok := c.Clouds[cloud]; !ok {
+		c.Clouds[cloud] = make(map[Name]cacheEntry)
+	}
```

Comment:
This isn't needed?  I suggest doing this in Set instead.

```suggestion
```

Reply:
More clearly: the first expiration check is only needed for Get (which doesn't need this if block) and this if block is only needed for the Set call (which doesn't need the expiration check).  My suggestion was more along the lines of not combining the two things..

Reply:
yeah, just realized that, working on it already